python_ga/mutation.py

```diff
@@ -0,0 +1,130 @@
+    :param config:
+    :return:
+    """
+    interval_in_indices = int(2 * config['time_interval_hours'])
+
+    mutation_config = config['mutations']['mutate_dose_value']
+    max_dose_value = config['max_dose_value']
+
+    max_dose = mutation_config['max_value']
+    min_dose = mutation_config['min_value']
+    step = mutation_config['step_value']
+    for i, genome in enumerate(population):
+        for gene_idx in range(len(genome)):
+            p = np.random.uniform()
+            if p < mutation_config['mut_prob']:
+                limit_value = min(max_dose, max_dose - sum(genome))
+                if limit_value > max_dose_value:
+                    limit_value = max_dose_value
+                if limit_value > min_dose:
+                    new_dose_value = np.random.randint(0, int(round((limit_value - min_dose) / step))) * step + min_dose
+                    genome[gene_idx] = new_dose_value
+                elif limit_value > 0:
+                    genome[gene_idx] = min_dose
+                    break
+                break
+        population[i] = refine_genome_around_cross_point_to_time_constraint(
+            genome=genome, interval_in_indices=interval_in_indices)
+    return population
+
+
+def mutate_time_value(population, config):
+    """
+    Mutacja zmian losowych czasu podania dawki
+    :param population:
+    :param config:
+    :return:
+    """
+    interval_in_indices = int(2 * config['time_interval_hours'])
+
+    mutation_config = config['mutations']['mutate_time_value']
+    max_dose_value = config['max_dose_value']
+
+    for i, genome in enumerate(population):
+        for gene_idx in range(len(genome)):
+            p = np.random.uniform()
+            if p < mutation_config['mut_prob']:
+                if genome[gene_idx] > 0:
+                    new_dose_time = np.random.randint(len(genome))
+                    new_dose_value = genome[new_dose_time] + genome[gene_idx]
+                    dose_remainder = 0
+                    if new_dose_value > max_dose_value:
+                        dose_remainder = new_dose_value - max_dose_value
+                        new_dose_value = max_dose_value
+
+                    genome[new_dose_time] = new_dose_value
+                    genome[gene_idx] = dose_remainder
+        population[i] = refine_genome_around_cross_point_to_time_constraint(
+            genome=genome, interval_in_indices=interval_in_indices)
+    return population
+
+
+def mutate_swap(population, config):
+    """
+    Mutacja wymiany: badany jest każdy gen w genomie wszystkich osobników w populacji. Jeżeli wylosowana wartość p jest
+    większa niż określone prawdopodobieństwo mutacji, wybrany gen jest zamieniamy miejscami z losowo wybranym innym genem
+    :param population:      list
+    :param config:          dict
+    :return: population:    list
+    """
+    interval_in_indices = int(2 * config['time_interval_hours'])
+
+    mutation_config = config['mutations']['mutate_swap']
```

Comment:
lines 200-208 from source are rewritten to 84-91 here, to the same iteration as in `mutate_time_value` and `mutate_dose_value`

algorithm logic not changed